src/main/kotlin/com/jetbrains/micropython/repl/ToolWindowReplTab.kt

```diff
@@ -0,0 +1,120 @@
+package com.jetbrains.micropython.repl
+
+import com.intellij.icons.AllIcons
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.actionSystem.ActionManager
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.actionSystem.DefaultActionGroup
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.application.ModalityState
+import com.intellij.openapi.project.DumbAware
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.Disposer
+import com.jediterm.terminal.TtyConnector
+import com.jetbrains.micropython.settings.MicroPythonDevicesConfiguration
+import org.jetbrains.plugins.terminal.JBTerminalSystemSettingsProvider
+import org.jetbrains.plugins.terminal.ShellTerminalWidget
+import java.awt.BorderLayout
+import javax.swing.JPanel
+
+class ToolWindowReplTab(val project: Project, parent: Disposable) : CommsEventObserver {
+    val deviceCommsManager = DeviceCommsManager.getInstance(project)
+    private val deviceConfiguration = MicroPythonDevicesConfiguration.getInstance(project)
+    val terminalWidget: ShellTerminalWidget
+
+    init {
+        val mySettingsProvider = JBTerminalSystemSettingsProvider()
+        terminalWidget = ShellTerminalWidget(project, mySettingsProvider, parent)
+
+        Disposer.register(terminalWidget, mySettingsProvider)
+
+        deviceCommsManager.registerObserver(this)
+        if (!deviceCommsManager.isRunning()) {
```

Comment:
These two calls would be sequential only if your communications manager is single-threaded (more specifically here: EDT-only). I guess it's the case, but could you please double-check if all "observer" (listener) events are delivered from the EDT?

Reply:
Will do.

Reply:
Sadly, calls to those functions can come from both EDT and application pooled threads (tested using `ApplicationManager.getApplication().assertIsDispatchThread()`). :(

I will implement some kind of synchronization to prevent possible race conditions.